proposals/0016-simd-application-fees.md

```diff
@@ -0,0 +1,438 @@
+decoded to calculate the total fees required for the transaction message. Then
+we verify that the fee-payer has a minimum balance of:
+`per-transaction base fees` + `prioritization fees` +
+`maximum application fees to be paid`
+
+If the payer does not have enough balance, the transaction is not scheduled
+and fails with an error `InsufficientFunds`. Consider this case the same as if
+the payer does not have enough funds to pay `other fees`.
+
+Before processing the message, we check if any loaded account has associated
+application fees. For all accounts with application fees, the fees paid should
+be greater than the required fees. In case of overpayment, the difference is
+stored in a variable and will be paid back to the payer in any case. In case the
+application fees are insufficiently paid or not paid then the transaction fails
+with an error `ApplicationFeesNotPaid`. The payer will pay just `other fees`.
+This failure is before the execution of the transaction.
+
+The application fees minus rebates and minus overpayment are transferred to
+the respecitve accounts lamport balance, from where the owner can collect them.
+
+### Configuration
+
+Application fees should be considered as set and forget kind of fees by dapp
+developers. They are not aimed to control congestion over the network instead
+they aim to reduce spamming.
+
+The ledger will need to store the amount of lamports required for every account.
+Hence changes should be constrained like any account write to the owner of the
+account. All programs will need to implement special instructions so that the
+authority of the accounts can sign to update application fees on the accounts.
+
+Setting the amount to `0` disables application fees. The maximum amount of
+application fees that can be set for an account will be limited to a constant
+amount of SOL (e.g. 20 SOL) so that accounts cannot become inaccessible through
+simple programming errors.
+
+### Rebate
+
+The owner of the account can issue a rebate of the application fees paid to
+write lock a specific account. Similar to the configuration this will need to be
+exposed through special instructions so that programs can implement custom
+authorization and delegate the decision to composing programs.
+
+Simple rebate schemes will verify merely signers, e.g an oracle preventing 3rd
+parties from write-locking their price feed. Dexes will need more complex rebate
+schemes based on instruction sysvar introspection.
+
+Rebate takes the amount of lamports to be rebated, and account on which rebate
+is issued as input. In case of multiple rebates from the same account only the
+highest amount of rebate will be taken into account. The rebated amount is
+always the minimum of rebate issued by the program and the application fees on
+the account. If program rebates `U64::MAX` it means all the application fees on
+the account are rebated. The rebate amount cannot be negative.
```

Comment:
This paragraph is a little confusing, I don't really understand what it means at all.

A "rebate", based on my understanding of the SIMD so far, is simply a manual lamport transfer by the program out of a writable account that received an account-write fee before the Message was executed back to the fee payer account (which I guess now needs to be explicitly passed into the instruction performing the rebate?)

Reply:
Yes, the rebate mechanism is not as simple as doing lamport transfers. We want to ensure on solana runtime side that in any case we do not rebate more than the application fees required to do the transactions. We also want to efficiently handle cases where multiple rebates are done on the same accounts.

Reply:
What is the exact mechanism behind "The owner of the account can issue a rebate of the application fees paid to
write lock a specific account"? A new syscall?

Reply:
it's flexible, we initially proposed a new instruction but then decided to leave it open for later discussions if this should be solved as a syscall, a new instruction or possibly both. in the current runtime CPI has a lot of overhead, especially when passing a large account, but instruction on the same program that implements the PayFees instruction would be easier to communicate to developers. wdyt?

Reply:
I see. It'd be nice to be super clear about what's still open for discussion in general, to prevent people from thinking that a SIMD is ready to land prematurely.

This seems like the tricky part in general with this proposal. How does it work if an account with an associated program-rebatable account write fee is passed into multiple instructions in the transaction? Is each instruction (including CPIs) charged the write fee? Seems like a footgun for the program otherwise, as now any rebate instruction needs to carefully consider the entire transaction to figure the correct amount. This is pretty tedious code to write, correctly. 

But regardless, I feel like the instruction performing the rebate probably just needs to receive the account that should receive the rebate as a normal instruction account (which the transaction creator most likely sets as the fee payer account)

Reply:
Yes we agree, as we are working on this proposal for long time we may have assumed that few things are obvious. 
The program relatable account write fees (PRAW fees) will be valid for whole transaction. So even if multiple instructions are using them they should be paid once per transaction.

The aim of the proposal is always to have trade off between usability for the dapps and changes that should be done in solana validator client. In this case we know that the payer is going to pay the application fees then why to add an additional complexity to use the rebate instruction for the dapps, this case will be easily handled by solana runtime reverting all rebates back to the payer. There will be more edge cases and complexity even on solana validator client  if we add account that should receive the rabate as normal account.

Reply:
If the write account fee is transaction wide, what's the mechanism for programs with multiple instructions in a single transaction that may wish to rebate the fee to not conflict with each other, and specifically return too much back?